templates/cdk.master.auth-webhook.py

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python3
+
+import csv
+import json
+import logging
+from base64 import b64decode
+from flask import Flask, request, jsonify
+from pathlib import Path
+from subprocess import check_output
+app = Flask(__name__)
+
+
+def kubectl(*args):
+    '''Run a kubectl cli command with a config file.
+
+    Returns stdout and throws an error if the command fails.
+    '''
+    command = ['kubectl', '--kubeconfig=/root/.kube/config'] + list(args)
+    return check_output(command)
+
+
+def check_known_tokens(token_review):
+    '''Populate user info if token is found in known_tokens.csv.'''
+    app.logger.info('Checking token')
+    token_to_check = token_review['spec']['token']
+
+    csv_fields = ['token', 'username', 'user', 'groups']
+    known_tokens = Path('/root/cdk/known_tokens.csv')
+    try:
+        with known_tokens.open('r') as f:
+            data_by_token = {r['token']: r for r in csv.DictReader(f, csv_fields)}
+    except FileNotFoundError:
+        data_by_token = {}
+
+    if token_to_check in data_by_token:
+        record = data_by_token[token_to_check]
+        token_review['status'] = {
+            'authenticated': True,
+            'user': {
+                'username': record['username'],
+                'uid': record['user'],
+                'groups': record['groups'].split(','),
+            }
+        }
+        return True
+    return False
+
+
+def check_secrets(token_review):
+    '''Populate user info if token is found in k8s secrets.'''
+    app.logger.info('Checking secret')
+    token_to_check = token_review['spec']['token']
+
+    output = kubectl('get', 'secrets', '-o', 'json').decode('UTF-8')
+    secrets = json.loads(output)
+    if 'items' in secrets:
+        for secret in secrets['items']:
+            try:
+                data_b64 = secret['data']
+                token_b64 = data_b64['password'].encode('UTF-8')
+            except (KeyError, TypeError):
+                continue
+
+            token = b64decode(token_b64).decode('UTF-8')
+            if token == token_to_check:
+                username_b64 = data_b64['username'].encode('UTF-8')
+                user_b64 = data_b64['user'].encode('UTF-8')
+                groups_b64 = data_b64['groups'].encode('UTF-8') \
+                    if 'groups' in data_b64 else b''
+
+                token_review['status'] = {
+                    'authenticated': True,
+                    'user': {
+                        'username': b64decode(username_b64).decode('UTF-8'),
+                        'uid': b64decode(user_b64).decode('UTF-8'),
+                        'groups': b64decode(groups_b64).decode('UTF-8').split(','),
+                    }
+                }
+                return True
+    return False
+
+
+def check_aws_iam(token_review):
+    app.logger.info('Checking AWS')
+    return False
+
+
+def check_keystone(token_review):
+    app.logger.info('Checking Keystone')
+    return False
+
+
+@app.route('/', methods=['POST'])
```

Comment:
You may want to add some versioning in the url so we can assist upgrades, beta testing etc